### PR TITLE
Restyling/form entry

### PR DESCRIPTION
**Pull request by dcluna**
New form entry screen, with new progress bar enabled by default.

### CODE REVIEW

File: app/src/org/commcare/android/view/ViewUtil.java [hunk 1]
```diff
@@ -81,4 +94,34 @@ public static Bitmap inflateDisplayImage(Context context, String jrUri) {
...
+        }
+    }
+    
+    /**
+     * Sets the background on a view to the provided drawable while retaining the padding
+     * of the original view (regardless of whether the provided drawable has its own padding)
+     * 
+     * @param v The view whose background will be updated
+     * @param background A background drawable (can be null to clear the background)
+     */
```
Comment (phillipm) [inline app/src/org/commcare/android/view/ViewUtil.java]: Nice!

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 2]
```diff
@@ -543,6 +550,38 @@ public void onClick(View v) {
         }
     }
 
+    private void setClickListenersForEverything() {
```
Comment (phillipm) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Unused and doesn't have a comment

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 3]
```diff
@@ -1059,46 +1107,45 @@ private void updateFloatingLabels(View currentView) {
...
-                layout.addView(right);
+
                 parent.addView(layout);
             } else {
                 largeLabels.add(smallLabels.get(i));
             }
         }
         for(int i = 0 ; i < largeLabels.size(); ++i ) {
-            TextView view = (TextView)View.inflate(this, R.layout.component_floating_label, null);
+            final TextView view = (TextView)View.inflate(this, R.layout.component_floating_label, null);
```
Comment (phillipm) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: +1 for adding finals

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 4]
```diff
@@ -1796,9 +1843,25 @@ public void showView(View next, AnimationType from, boolean animateLastView) {
         mViewPane.addView(mCurrentView, lp);
 
         mCurrentView.startAnimation(mInAnimation);
-        if (mCurrentView instanceof ODKView)
+        
+        if (mCurrentView instanceof ODKView) {
             ((ODKView) mCurrentView).setFocus(this);
-        else {
+            
+            String groupLabelText = ((ODKView) mCurrentView).getGroupLabel();
```
Comment (phillipm) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Whenever possible try to define a variable as close as possible to its usage (so down near line 1859). Unless it is a class level variable, which should go at the top of the class, contrary to what Clayton may say :p

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 5]
```diff
@@ -1796,9 +1843,25 @@ public void showView(View next, AnimationType from, boolean animateLastView) {
...
+
+            FrameLayout header = (FrameLayout)mCurrentView.findViewById(R.id.form_entry_header);
+
+            TextView groupLabel = ((TextView)header.findViewById(R.id.form_entry_group_label));
+
+            header.setVisibility(View.GONE);
+            groupLabel.setVisibility(View.GONE);
+
+            if(groupLabelText != "") {
```
Comment (phillipm) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Never use ==/!= with Strings (also always use a linter, it will warn you of this). 
This shouldn't work unless you change it to !"".equals(groupLabelText)

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 6]
```diff
@@ -2857,7 +2920,11 @@ public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float ve
             if (velocityX > 0) {
                 showPreviousView();
             } else {
-                showNextView();
+                int event = mFormController.getEvent(mFormController.getNextFormIndex(mFormController.getFormIndex(), true));
+                boolean navBar = PreferencesActivity.getProgressBarMode(this).useNavigationBar();
```
Comment (phillipm) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: I like to give boolean variables names that sound like questions, like 'hasNavBar' or 'usingNavBar'. This helps (whatever small amount) when reading complex boolean expressions.

File: app/src/org/odk/collect/android/views/ODKView.java [hunk 7]
```diff
@@ -66,7 +66,10 @@
     private int mViewBannerCount = 0;
 
     private boolean mProgressEnabled;
+    
+    String mGroupLabel;
 
+    private static boolean SEPERATORS_ENABLED = false;
```
Comment (phillipm) [inline app/src/org/odk/collect/android/views/ODKView.java]: final?

Comment (phillipm) [inline app/src/org/odk/collect/android/views/ODKView.java]: comment informing us as to what this effects

File: app/src/org/odk/collect/android/views/ODKView.java [hunk 8]
```diff
@@ -312,7 +322,7 @@ public void updateProgressBar(int progress, int max) {
     /**
      * // * Add a TextView containing the hierarchy of groups to which the question belongs. //
```
Comment (phillipm) [inline app/src/org/odk/collect/android/views/ODKView.java]: This method's docs are now out-of-date.

File: app/src/org/odk/collect/android/widgets/QuestionWidget.java [hunk 9]
```diff
@@ -80,6 +84,9 @@ public QuestionWidget(Context context, FormEntryPrompt p) {
     public QuestionWidget(Context context, FormEntryPrompt p, WidgetChangedListener w){
         super(context);
 
+        //this is pretty sketch but is the only way to make the required background to work trivially for now
```
Comment (phillipm) [inline app/src/org/odk/collect/android/widgets/QuestionWidget.java]: hahaha, looks like 'sketch' is required vocabulary for the mobile team

File: app/src/org/odk/collect/android/widgets/QuestionWidget.java [hunk 10]
```diff
@@ -107,12 +114,15 @@ public void onClick(View v) {
 
         setOrientation(LinearLayout.VERTICAL);
         setGravity(Gravity.TOP);
-        setPadding(0, 7, 0, 0);
+        
+        //TODO: This whole view should probably be inflated somehow
```
Comment (phillipm) [inline app/src/org/odk/collect/android/widgets/QuestionWidget.java]: I like the idea of adding your initials right after the TODO. This makes it easy for one to search for their TODOs so they can easily revisit them.

File: app/src/org/odk/collect/android/widgets/QuestionWidget.java [hunk 11]
```diff
@@ -202,15 +221,27 @@ public void notifyOnScreen(String text, boolean strong){
...
+            // We could accomplish the same thing by having two backgrounds, one for the widget as a whole, and 
+            // one for the internals (or splitting up the layout), but this'll do for now 
+            this.toastView = View.inflate(this.getContext(), R.layout.toast_view_modern, this).findViewById(R.id.toast_view_root);
+
+//            
+//            LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) toastView.getLayoutParams();
+//            int margin = -1 * this.getResources().getDimensionPixelSize(R.dimen.question_widget_side_padding);
+//
+//            params.setMargins(margin, 0,margin,0);
+//            this.addView(toastView, params);
```
Comment (phillipm) [inline app/src/org/odk/collect/android/widgets/QuestionWidget.java]: Why is this commented code here?

File: app/src/org/odk/collect/android/widgets/QuestionWidget.java [hunk 12]
```diff
@@ -406,11 +434,15 @@ protected void addQuestionText(final FormEntryPrompt p) {
         addView(mediaLayout, mLayout);
     }
 
+    private void fireHelpText(FormEntryPrompt prompt) {
```
Comment (phillipm) [inline app/src/org/odk/collect/android/widgets/QuestionWidget.java]: This method doesn't seem to ever be used.

**Comment by phillipm** [comment]
Noting that I made a review pass on the Java portions a few weeks.

File: app/res/drawable/bubble_invalid_modern.xml [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+               <!--android:bottomRightRadius="12dp"-->
+               <!--android:bottomLeftRadius="12dp"-->
+               <!--android:topLeftRadius="12dp"-->
+               <!--android:topRightRadius="12dp"-->
+           	   <!--/>-->
+        </shape>
+    </item>
+
+
+    <item android:left="8dp" android:top="8dp" android:bottom="8dp" android:right="8dp">
```
Comment (ctsims) [inline app/res/drawable/bubble_invalid_modern.xml]: naked padding

File: app/res/drawable/label_floating_bad.xml [hunk 14]
```diff
@@ -1,10 +1,13 @@
...
-    <item>
-        <shape android:shape="rectangle"> 
-            <solid android:color="@color/solid_dark_red"/>
+     <item>
+        <shape android:shape="rectangle">
+            <solid android:color="@color/cc_attention_negative_color"/>
          </shape>
     </item>
-
+    <item android:left="8dp" android:right="8dp">
```
Comment (ctsims) [inline app/res/drawable/label_floating_bad.xml]: These all have the same dimensions, right? would be good to style them somehow so these drawables are consistent without needing to keep them in sync

File: app/res/layout/odkview_layout.xml [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+          android:paddingLeft="16dp"
+          android:paddingRight="16dp"
+            />
+    </FrameLayout>
+    <LinearLayout
+        android:id="@+id/odkview_layout"
+        android:orientation="vertical"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:paddingTop="7px"
```
Comment (ctsims) [inline app/res/layout/odkview_layout.xml]: is 7px ok? This should be 8dp, right?

File: app/res/layout/title_text_view_modern.xml [hunk 16]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+          android:layout_width="match_parent"
+          android:layout_height="match_parent"
+          android:paddingTop="16dp"
```
Comment (ctsims) [inline app/res/layout/title_text_view_modern.xml]: should likely be in a style

File: app/res/layout/toast_view_modern.xml [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        android:paddingLeft="8dp"
+    	android:paddingRight="8dp">
+
+            <TextView
+                android:id="@+id/message"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                android:gravity="center"
+                android:paddingBottom="16dp"
```
Comment (ctsims) [inline app/res/layout/toast_view_modern.xml]: same dimensions as the textview above. Style?

File: app/src/org/commcare/android/view/ViewUtil.java [hunk 18]
```diff
@@ -26,13 +39,13 @@
  * @author ctsims
  *
  */
-public class ViewUtil {
+public final class ViewUtil {
```
Comment (ctsims) [inline app/src/org/commcare/android/view/ViewUtil.java]: what's the purpose of these changes?

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 19]
```diff
@@ -241,21 +246,21 @@
...
 
         // See if this form needs GPS to be turned on
         mNoGPSReceiver = new BroadcastReceiver() {
             @Override
-            public void onReceive(Context context, Intent intent) {
+            public void onReceive(final Context context, final Intent intent) {
                 context.removeStickyBroadcast(intent);
-                LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
-                Set<String> providers = GeoUtils.evaluateProviders(manager);
+                final LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
```
Comment (ctsims) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: These 'final' changes are making the diff hard to read. Are they stylistic or functional?

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 20]
```diff
@@ -543,6 +551,41 @@ public void onClick(View v) {
         }
     }
 
+    /**
+     * Debug method to toast a view's ID whenever it is clicked.
+     */
+    private void setClickListenersForEverything() {
```
Comment (ctsims) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: This seems potentially useful, but I'd put it into a view utility/instrumentation class statically

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java (hunk 19 above)
Comment (dcluna) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Stylistic, mostly based on linter suggestions. I'm removing them in my next commit to help visualization.

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java (hunk 6 above)
Comment (dcluna) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Updated to hasNavBar

File: app/src/org/odk/collect/android/views/ODKView.java (hunk 8 above)
Comment (dcluna) [inline app/src/org/odk/collect/android/views/ODKView.java]: Updated

File: app/src/org/odk/collect/android/widgets/QuestionWidget.java (hunk 12 above)
Comment (dcluna) [inline app/src/org/odk/collect/android/widgets/QuestionWidget.java]: It is used in line 157 of the same class
